File: apps/aperture-ios/Aperture/DesignSystem/Format.swift
// Formatting helpers — iOS twin of src/lib/format.ts. Consistent number/percent
// rendering with tabular figures.
import Foundation
import SwiftUI

enum Fmt {
    static func pct(_ v: Double, _ dp: Int = 2) -> String { String(format: "%.\(dp)f%%", v * 100) }
    static func pctSigned(_ v: Double, _ dp: Int = 2) -> String {
        (v >= 0 ? "+" : "") + String(format: "%.\(dp)f%%", v * 100)
    }
    static func signed(_ v: Double, _ dp: Int = 2) -> String {
        (v >= 0 ? "+" : "") + String(format: "%.\(dp)f", v)
    }
    static func prob(_ v: Double?) -> String { v == nil ? "—" : String(format: "%.2f", v!) }
    static func bps(_ v: Double) -> String { (v >= 0 ? "+" : "") + String(format: "%.1f bps", v) }

    static func price(_ v: Double) -> String {
        v >= 1 ? num(v, 2) : num(v, 4)
    }
    static func num(_ v: Double, _ dp: Int = 2) -> String {
        let f = NumberFormatter()
        f.numberStyle = .decimal
        f.minimumFractionDigits = dp
        f.maximumFractionDigits = dp
        return f.string(from: NSNumber(value: v)) ?? String(format: "%.\(dp)f", v)
    }
    static func compact(_ v: Double, prefix: String = "$") -> String {
        let abs = Swift.abs(v), sign = v < 0 ? "-" : ""
        if abs >= 1e12 { return "\(sign)\(prefix)\(String(format: "%.2f", abs / 1e12))T" }
        if abs >= 1e9 { return "\(sign)\(prefix)\(String(format: "%.2f", abs / 1e9))B" }
        if abs >= 1e6 { return "\(sign)\(prefix)\(String(format: "%.2f", abs / 1e6))M" }
        if abs >= 1e3 { return "\(sign)\(prefix)\(String(format: "%.1f", abs / 1e3))K" }
        return "\(sign)\(prefix)\(String(format: "%.0f", abs))"
    }
    static func timeAgo(_ seconds: Int) -> String {
        if seconds < 60 { return "\(seconds)s ago" }
        if seconds < 3600 { return "\(seconds / 60)m ago" }
        return "\(seconds / 3600)h ago"
    }
    static func sideLabel(_ side: Int) -> String { side > 0 ? "LONG" : side < 0 ? "SHORT" : "FLAT" }
}

extension Action {
    /// (fill color, text color) for the pill.
    var pillColors: (bg: Color, fg: Color) {
        switch self {
        case .BUY: return (Tok.buySoft, Tok.buy)
        case .SELL: return (Tok.sellSoft, Tok.sell)
        case .WATCH: return (Tok.watchSoft, Tok.watch)
        case .MODEL_REQUIRED, .NO_DATA: return (Tok.neutralSoft, Tok.neutral)
        }
    }
}

File: apps/aperture-web/src/lib/format.ts
// Formatting helpers — consistent number/currency/percent rendering.

export const fmtUsd = (v: number, dp = 2): string =>
  v.toLocaleString('en-US', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: dp,
    maximumFractionDigits: dp,
  });

export const fmtCompact = (v: number, prefix = '$'): string => {
  const abs = Math.abs(v);
  const sign = v < 0 ? '-' : '';
  if (abs >= 1e12) return `${sign}${prefix}${(abs / 1e12).toFixed(2)}T`;
  if (abs >= 1e9) return `${sign}${prefix}${(abs / 1e9).toFixed(2)}B`;
  if (abs >= 1e6) return `${sign}${prefix}${(abs / 1e6).toFixed(2)}M`;
  if (abs >= 1e3) return `${sign}${prefix}${(abs / 1e3).toFixed(1)}K`;
  return `${sign}${prefix}${abs.toFixed(0)}`;
};

export const fmtNum = (v: number, dp = 2): string =>
  v.toLocaleString('en-US', { minimumFractionDigits: dp, maximumFractionDigits: dp });

export const fmtPrice = (v: number): string => {
  if (v >= 1000) return fmtNum(v, 2);
  if (v >= 1) return fmtNum(v, 2);
  return fmtNum(v, 4);
};

export const fmtPct = (v: number, dp = 2): string => `${(v * 100).toFixed(dp)}%`;

export const fmtPctSigned = (v: number, dp = 2): string =>
  `${v >= 0 ? '+' : ''}${(v * 100).toFixed(dp)}%`;

export const fmtSigned = (v: number, dp = 2): string => `${v >= 0 ? '+' : ''}${v.toFixed(dp)}`;

export const fmtBps = (v: number): string => `${v >= 0 ? '+' : ''}${v.toFixed(1)} bps`;

export const fmtProb = (v: number | null): string => (v == null ? '—' : v.toFixed(2));

export const signClass = (v: number): string => (v > 0 ? 'pos' : v < 0 ? 'neg' : '');

export const sideLabel = (side: number): string =>
  side > 0 ? 'LONG' : side < 0 ? 'SHORT' : 'FLAT';

export const fmtTimeAgo = (seconds: number): string => {
  if (seconds < 60) return `${Math.round(seconds)}s ago`;
  if (seconds < 3600) return `${Math.round(seconds / 60)}m ago`;
  return `${Math.round(seconds / 3600)}h ago`;
};
